src/ProductConstructionService/ProductConstructionService.FeedCleaner/FeedCleanerJob.cs

```diff
@@ -0,0 +1,161 @@
+        Microsoft.Extensions.Logging.ILogger<FeedCleanerJob> logger)
+    {
+        _serviceProvider = serviceProvider;
+        _azureDevOpsClient = azureDevOpsClient;
+        _options = options;
+        _logger = logger;
+    }
+
+    private FeedCleanerOptions Options => _options.Value;
+
+    public async Task CleanManagedFeedsAsync()
+    {
+        if (!Options.Enabled)
+        {
+            _logger.LogInformation("Feed cleaner service is disabled in this environment");
+            return;
+        }
+
+        _logger.LogInformation("Loading packages in release feeds...");
+
+        PackagesInReleaseFeeds packagesInReleaseFeeds = await GetPackagesForReleaseFeedsAsync();
+
+        _logger.LogInformation("Loaded {versionCount} versions of {packageCount} packages from {feedCount} feeds",
+            packagesInReleaseFeeds.Sum(feed => feed.Value.Sum(package => package.Value.Count)),
+            packagesInReleaseFeeds.Sum(feed => feed.Value.Keys.Count),
+            packagesInReleaseFeeds.Keys.Count);
+
+        foreach (var azdoAccount in Options.AzdoAccounts)
+        {
+            _logger.LogInformation("Processing feeds for {account}...", azdoAccount);
+
+            List<AzureDevOpsFeed> allFeeds;
+            try
+            {
+                allFeeds = await _azureDevOpsClient.GetFeedsAsync(azdoAccount);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Failed to get feeds for account {azdoAccount}", azdoAccount);
+                continue;
+            }
+
+            List<AzureDevOpsFeed> managedFeeds = allFeeds
+                .Where(f => Regex.IsMatch(f.Name, FeedConstants.MaestroManagedFeedNamePattern))
+                .Shuffle()
+                .ToList();
+
+            _logger.LogInformation("Found {totalCount} feeds for {account}. Will process {count} matching feeds",
+                allFeeds.Count,
+                managedFeeds.Count,
+                azdoAccount);
+
+            int feedsCleaned = 0;
+
+            Parallel.ForEach(
+                managedFeeds,
+                new ParallelOptions
+                {
+                    MaxDegreeOfParallelism = 5
+                },
+                async (AzureDevOpsFeed feed) =>
+                {
+                    try
+                    {
+                        using var scope = _serviceProvider.CreateScope();
+                        using var feedCleaner = scope.ServiceProvider.GetRequiredService<FeedCleaner>();
+                        await feedCleaner.CleanFeedAsync(feed, packagesInReleaseFeeds);
+                        Interlocked.Increment(ref feedsCleaned);
+                    }
+                    catch (Exception e)
+                    {
+                        _logger.LogError(e, "Failed to clean feed {feed}", feed.Name);
+                    }
+                });
+
+            _logger.Log(
+                feedsCleaned != managedFeeds.Count ? LogLevel.Warning : LogLevel.Error,
```

Comment:
Are these log levels correct? If it deleted all which was asked to delete, then shouldn't it be Information only? And error or warning if something was not properly deleted

Reply:
```suggestion
                feedsCleaned != managedFeeds.Count ? LogLevel.Warning : LogLevel.Information,
```

Good catch